Dupes cleanup Procedure on db.

ricdezen opened the pull request:
Added stored procedure code for future reference.
Deletes duplicate rows while checking that no unique row got deleted.

ricdezen committed:
feat(cleanup_dupes.sql)

Added stored procedure code for future reference.
Deletes duplicate rows while checking that no unique row got deleted.

CremaLuca commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953&bid=19489274)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+		RAISE WARNING 'Some meaningful data got deleted, rolling back.';
+		ROLLBACK;
+		RETURN;
+	ELSE
```
Codacy found an issue: [syntax error at or near "ELSE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "ELSE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	
+	RAISE NOTICE 'Rows after cleanup %.', (SELECT COUNT(*) FROM atoms_b);
+END;
+$$;
```
Codacy found an issue: [unterminated dollar-quoted string at or near "$$;](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [unterminated dollar-quoted string at or near "$$;](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	
+	c_after = (SELECT COUNT(DISTINCT data_json) FROM atoms_b);
+	-- If some meaningful data was deleted, rollback.
+	IF c_after < c_before THEN
```
Codacy found an issue: [syntax error at or near "IF"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "IF"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+    c_before bigint;
+	c_after bigint;
+BEGIN
+	RAISE NOTICE 'Rows before cleanup %.', (SELECT COUNT(*) FROM atoms_b);
```
Codacy found an issue: [syntax error at or near "RAISE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "RAISE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	c_after bigint;
+BEGIN
+	RAISE NOTICE 'Rows before cleanup %.', (SELECT COUNT(*) FROM atoms_b);
+    c_before = (SELECT COUNT(DISTINCT data_json) FROM atoms_b);
```
Codacy found an issue: [syntax error at or near "c_before"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "c_before"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	IF c_after < c_before THEN
+		RAISE WARNING 'Some meaningful data got deleted, rolling back.';
+		ROLLBACK;
+		RETURN;
```
Codacy found an issue: [syntax error at or near "RETURN"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "RETURN"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+CREATE OR REPLACE PROCEDURE cleanup_dupes()
```
Codacy found an issue: [syntax error at or near "PROCEDURE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "PROCEDURE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+AS $$
+DECLARE
+    c_before bigint;
+	c_after bigint;
```
Codacy found an issue: [syntax error at or near "c_after"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "c_after"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	WHERE a.id < b.id
+    AND a.data_json = b.data_json;
+	
+	c_after = (SELECT COUNT(DISTINCT data_json) FROM atoms_b);
```
Codacy found an issue: [syntax error at or near "c_after"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "c_after"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+		COMMIT;
+	END IF;
+	
+	RAISE NOTICE 'Rows after cleanup %.', (SELECT COUNT(*) FROM atoms_b);
```
Codacy found an issue: [syntax error at or near "RAISE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "RAISE"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+		RETURN;
+	ELSE
+		COMMIT;
+	END IF;
```
Codacy found an issue: [syntax error at or near "IF"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca replied:
Codacy found an issue: [syntax error at or near "IF"](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953)

CremaLuca commented on misc/cleanup_dupes.sql:
```diff
@@ -0,0 +1,29 @@
+	-- Delete dupes
+	DELETE FROM
+    atoms_b a USING atoms_b b
+	WHERE a.id < b.id
```
Why not something more complex? What if one of the atoms is contained in another one?
[This trello board](https://trello.com/c/57EQA2xD) contains a few queries.

```
WHERE a.id < b.id
   AND b.data_json <@ a.data_json;
```

By checking "is contained" you could be removing distinct values (if the two atoms are not equal) and the distinct count of values check wouldn't work.

IDK I'm fine with this cleanup method too, just consider the is contained check and let me know.

Commit:
feat(config.py):
+ Added method to build a callable for a specific file.

Misc:
+ Added sql_alchemy adapter draft
+ Added "local.json" to .gitignore, file used when testing in a local database

Commit:
Merge branch 'hotfix/db-cleanup' of https://github.com/OTRI-Unipd/OTRI into hotfix/db-cleanup

CremaLuca commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- sqlalchemy_messaround.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=5944953&bid=19489274)
